### PR TITLE
[stable/mattermost-team-edition] Configure everything in one go

### LINKED ISSUE #8008: [stable/mattermost-team-edition] Passthrough configuration
## Feature request
It would  be great to be able to pass through arbitrary configuration that makes sense to the mattermost server. This is certainly not easy with Helm 2 where everything you have entered in anything except within `.Values` is treated as text.

Perhaps you could move the default values declared in `config.tpl` to values.yaml, like auth.gitlab already is. But, you need also an intermediary step...

Look at this, from `config.tpl` for example, notice how you set certain fields based on externalDB.enabled:

```yaml
    "SqlSettings": {
        {{ if .Values.externalDB.enabled }}
        "DriverName": "{{ .Values.externalDB.externalDriverType }}",
        "DataSource": "{{ .Values.externalDB.externalConnectionString }}",
        {{ else }}
        "DriverName": "mysql",
        "DataSource": "{{ .Values.mysql.mysqlUser }}:{{ .Values.mysql.mysqlPassword }}@tcp({{ .Release.Name }}-mysql:3306)/{{ .Values.mysql.mysqlDatabase }}?charset=utf8mb4,utf8&readTimeout=30s&writeTimeout=30s",
        {{ end }}
```

So, you want to adjust the value based on something, or not allow the user to configure anything, etc. How would you do this? Well... Consider how my friend @minrk does it in the following essential PR for the Helm chart I'm working on as well: https://github.com/jupyterhub/zero-to-jupyterhub-k8s/pull/941/files#diff-96a1791deadeccb98a6f334d195c73a7R23

By importing the values from `values.yaml`, storing them in a Helm template variable, and manipulating them, and then rendering them as JSON!

---

The benefits of doing it like this can be immense. You would not need to duplicate documentation, but instead refer to the mattermost-server documentation. I would also have been able to configure other hardcoded settings which I now can't like `EmailSettings.EnableSignInWithEmail` etc.

If you feel like a video chat about this, I'm happy to help you discuss this with you.

/cc: @cpanato @jwilander (because Chart.yaml indicates you to be the maintainers)

### PR DESCRIPTION
**Pull request by consideRatio**
## Background
It can be daunting to maintain a Chart that passes configuration to a container by adding yaml to a configmap. The PR fixes #8008. This could be vital to the sustainability of maintaining a chart like this, as there is simply a lot of configuration to pass through and it can be hard to keep up to date.

## About changes
* I moved all the config.json content from the configmap to `values.yaml` under `configJSON` because reading them from there makes them remain a map, rather than a big blob of text. This allows us to override parts etc.
* I ensured all previous values that so far has been configured to write to `config.json` were overriding things in in a copy of `.Values.configJSON` instead. The modified copy of `configJSON` were then written to the field `config.json` that becomes mounted to the mattermost server pod.
* As sensitive information can be stored in this config, I made the ConfigMap become a Secret instead.

## Breaking change
I make some breaking changes, but I also make sure to print the config changes required by the user automatically. So for example `auth.gitlab` is deprecated hard in favor of `configJSON.GitLabSettings` that does the same thing and matches the typical "config.json" file.

So for example, if a user tries to install this version but had declared something in `auth.gitlab` they will fail to render the templates and instead get a message about needing to use `configJSON.GitLabSettings` from now on. See link in the paragraph below.

To summarize, the following helm configuration values have been deprecated, and if you would use them you would receive an error on `helm upgrade` before `tiller` have started making changes so the upgrade will fail before any changes has been made to the deployment. And again, to see the actual deprecation failure output, see [this comment below in the PR](https://github.com/helm/charts/pull/8019#issuecomment-453820463).

- auth.gitlab
- config
  - siteUrl
  - siteName
  - enableSignUpWithEmail
  - feedbackName
  - feedbackEmail
  - smtpUsername
  - smtpPassword
  - smtpUsername
  - smtpPassword
  - smtpServer
  - smtpPort
  - smtpConnection

/assign @cpanato

### CODE REVIEW

**Comment by consideRatio** [comment]
@minrk gave me valuable input about this PR.

What I got away with was:
1. We may want to make a merge with the default configuration on the mattermost-server binary to avoid hardcoding a default config in here.
2. We should not have `defaultConfig` + `extraConfig` but simply `config` and allow the user to override that, we don't need defaultConfig + extraConfig or similar! Helm will merge it for us.

Note `config` should perhaps be referred to as `boilerplateConfig` or similar since this PR modifies it with custom Helm chart values for backward compatibility and simplicity. All fields currently assigned the value PLACEHOLDER are such fields.

---

Hmmm, okay the drawback of doing 2. would be that we end up with another kind of sequence of configuration updates.

**Current order of overrides in this PR:**
- `defaultConfig` (aka. `boilerplateConfig`) gets updated by custom logic for custom chart values like this:
  ```yaml
  {{- if .Values.externalDB.enabled }}
  {{- $_ := set $c.SqlSettings "DriverName" (.Values.externalDB.externalDriverType) }}
  {{- $_ := set $c.SqlSettings "DataSource" (.Values.externalDB.externalConnectionString) }}
  {{- else }}
  ```
- That config is then overridden with `extraConfig`

**If we implement `2.` we would have:** the same, but the order of updates would change. `extraConfig` would be applied first.

Not certain what is best. But I think I like the second version better.

**Comment by cpanato** [comment]
no rush @consideRatio thanks for all

**Comment by consideRatio** [comment]
I'll get to do this ... !

**Comment by consideRatio** [comment]
# Update Jan

With these changes, anyone knowing how to setup mattermost using a `config.json` will be able to configure this deployment as well through `.Values.configJSON`. Like this, any configurable option in mattermost-server will be configurable, even though this chart has not added a specific option for it, as compared to before where we utilized a mostly hardcoded `config.tpl` file that overrode certain fields with specific values in `.Values`.

If this PR makes sense to you, I will try update documentation as well. This is needed as I deprecated all configuration in `.Values.config` and `auth.gitlab`. But on the upside, I did it quite gracefully I think. For example: if anyone runs an old configuration they will receive a copy-pastable "do this instead".

Note that the `mysql` and `externalDB` options remain, mainly because `mysql` will configure both the dependent `mysql` helm chart as well as this charts connection to the mysql server. These two fields, `mysql` and `externalDB` will overwrite configuration in `configJSON` and some confusion could arise... For this reason, I made it so that if the user attempts to configure `configJSON.SqlSettings.DataSource` directly for example, the user will fail on template render time with a message to utilize `mysql` or `externalDB` instead to accomplish the same thing.

---

I have taken inspiration from [@minrk's PR in the JupyterHub helm chart](https://github.com/jupyterhub/zero-to-jupyterhub-k8s/pull/941) and looked to [GitLab's helm chart for inspiration on deprecations](https://gitlab.com/charts/gitlab/blob/master/templates/_deprecations.tpl).

# Future tasks

- [ ] @cpanato @jwilander give feedback about if this PR's goal makes sense to you and what else might need doing on it except writing the documentation
- [ ] @consideRatio Test the new changes on my own deployment of Mattermost as well.
- [ ] @consideRatio Write documentation.

**Comment by consideRatio** [comment]
# Deployment success!

## First `helm upgrade`
I run the upgrade but it failed like it should, in the following way.

```
Error: UPGRADE FAILED: render error in "mattermost-team-edition/templates/NOTES.txt": template: mattermost-team-edition/templates/NOTES.txt:31:3: executing "mattermost-team-edition/templates/NOTES.txt" at <include "mattermost....>: error calling include: template: mattermost-team-edition/templates/_deprecations.tpl:43:104: executing "mattermost.deprecations" at <fail>: error calling fail: 

FAILURE DUE TO DEPRECATIONS:
----------------------------

configJSON:
  # auth.gitlab is deprecated, instead use:
  GitLabSettings:
    AuthEndpoint: https://gitlab.example.com/oauth/authorize
    Enable: true
    Id: REDACTED
    Scope: ""
    Secret: REDACTED
    TokenEndpoint: https://gitlab.example.com/oauth/token
    UserApiEndpoint: https://gitlab.example.com/api/v4/user
    

  # config.siteUrl is deprecated, instead use:
  ServiceSettings:
    SiteURL: "https://mattermost.example.com"

  # config.siteName is deprecated, instead use:
  TeamSettings:
    SiteName: "Mattermost"

  # config.enableSignUpWithEmail,
  # config.feedbackName,
  # config.feedbackEmail,
  # config.smtpUsername,
  # config.smtpPassword,
  # config.smtpUsername,
  # config.smtpPassword,
  # config.smtpServer,
  # config.smtpPort,
  # config.smtpConnection,
  # are all deprecated, instead use:
  EmailSettings:
    ConnectionSecurity: ""
    EnableSMTPAuth: true
    EnableSignUpWithEmail: true
    FeedbackEmail: REDACTED
    FeedbackName: Mattermost
    SMTPPassword: REDACTED
    SMTPPort: 587
    SMTPServer: smtp-relay.sendinblue.com
    SMTPUsername: REDACTED
    SendEmailNotifications: true
```

## Second `helm upgrade`
After commenting out my `config` entry in my provided helm values and copy pasting the `configJSON` entry provided by the output above I got the following warning instead.

```
WARNING:
--------

Every `helm upgrade` will generate a new set of keys unless it is set manually like this:

configJSON:
  EmailSettings:
    InviteSalt: CfGFN4CrJtQ1NKZc7rbiMPHncxyGjUjM
  FileSettings:
    PublicLinkSalt: wAU3GUC2LoRJVMwuBrBQNHiJORsqkOFF
  SqlSettings:
    AtRestEncryptKey: 3UonCdN4jLvXWomxd8ZCRrofSVvebpNU
```

## Third `helm upgrade`

I added the additional salts / keys from the second upgrade attempt (that succeeded with a warning) and things worked fine!

**Comment by consideRatio** [comment]
The ConfigMap we reference should really be a secret as it contains secrets.

**Comment by jwilander** [comment]
I'll defer to @cpanato for the actual review of the changes but the direction sounds good to me. I particularly like that chart users would be modifying the config as JSON like they would if they were not running using the helm chart. No need to conceptually map the settings from JSON to YAML

**Comment by consideRatio** [comment]
@cpanato I updated the initial description to reflect the evolution of the PR.

**Comment by cpanato** [comment]
@consideRatio need a rebase to fix the chart conflict

**Comment by consideRatio** [comment]
**Final tasks**:
- [ ] Assuming usage of hard deprecations, ie forcing failures in NOTES.txt if deprecated config solution is in use, I should increment the major instead of minor version according to SemVer, I think.
- [ ] Documentation update of changes
- [ ] Documentation completion of lacked fields like externalDB

File: stable/mattermost-team-edition/README.md [hunk 1]
```diff
@@ -90,6 +83,24 @@ Alternatively, a YAML file that specifies the values for the parameters can be p
...
+A basic example of a `.yaml` file with values that could be passed to the `helm`
+command with the `-f` or `--values` flag to get started.
+
+```yaml
+ingress:
+  enabled: true
+  hosts:
+    - mattermost.example.com
+
+configJSON:
```
Comment (cpanato) [inline stable/mattermost-team-edition/README.md:105]: the example you say you can use like a yaml, but in the values.yaml they are in json format. I'm a bit confused. can you clarify?

and thanks for this, looks cool

Comment (consideRatio) [reply stable/mattermost-team-edition/README.md:105]: Hi @cpanato !

We utilize .yaml files for the helm chart configuration, but as YAML allows JSON syntax as well to declare objects, we can within a .yaml file utilize JSON or YAML. In the end, however the object was defined, we will output it with JSON syntax in secret-config.yaml.

```
  # from within templates/secret-config.yaml
  # ...
  # $c is the configJSON object, which first has been merged with the user provided values, and
  # then copied and modified based on chart configuration such as `externalDB.enabled`
  # the "| toJson" part makes the output JSON no matter how it initially was declared
  config.json: {{ $c | toJson | b64enc }}
```

Comment (consideRatio) [reply stable/mattermost-team-edition/README.md:105]: So, no matter if .Values.configJSON was in YAML or pure JSON formatting, it is converted to JSON in the end.

**Comment by consideRatio** [comment]
@cpanato I understand that you may have a lot on your plate as I've seen you do work all over. Is there something I could help you with?

I care for this kind of changes as were essential for me to deploy and configure Mattermost myself, but also from experience of working on other helm charts I believe that these kinds of config passthroughs from helm to the actual binary that is run can reduce so much time invested maintaining and keeping the deployment up to date with features etc.

**Comment by consideRatio** [comment]
Rebased to account for Mattermost version updates and @ahmadalli's #11348 to avoid merge conflicts.

**Comment by cpanato** [comment]
@consideRatio First I need to apologize for not being responsive and we get merge this change. this is totally my fault.
I will review and make any changes if you allow me to do to and then we get this merge this week.
again sorry :(

**Comment by cpanato** [comment]
@consideRatio can we add some documentation in the readme as well? what happen if i already have a installed chart and upgrade to this new one? do i need to remove and reinstall?

**Comment by consideRatio** [comment]
@cpanato no worries!

Ah yes we should prepare the users for the migration, what will happen is documented in this comment right now: https://github.com/helm/charts/pull/8019#issuecomment-453820463

Should such documentation be within the readme, and or somewhere else?

In short the upgrade process is like this, if they try to make an upgrade to from 2.x.x to 3.x.x without reading anything about it, the upgrade will fail before any changes to the deployment has been made, and they are suggested with changes to the helm config they need to make.

Feel free to make changes to the PR!

## TODO
- [ ] Upgrade related documentation
- [ ] Decision: Should the configJSON field in `values.yaml` stay as JSON, or should it be YAML?
- [ ] Decision: Should the deprecation errors that suggest how to write the config then be presented as JSON as well perhaps?

File: stable/mattermost-team-edition/README.md (hunk 1 above)
Comment (consideRatio) [reply stable/mattermost-team-edition/README.md:105]: I figure we may want to decide on sticking to JSON or YAML throughout. JSON is a subset of YAML though so it is just a matter of establishing a convention as it will always really be interpreted as YAML (which may be pure JSON if we decide to use that as a convention for all things related to the config.json file that we generate).

**Comment by cpanato** [comment]
@consideRatio yes, the migration/upgrade should be documented in the readme in a specific section. this will be great! :)

**Comment by cpanato** [comment]
@consideRatio regarding this ` Should the configJSON field in values.yaml stay as JSON, or should it be YAML?` I dont have a strong opinion. 
I think we should keep what is easiest for the user.

**Comment by cpanato** [comment]
@consideRatio tests are passing, I also installed locally and is working :) so, I would say just the documentation and any other thing that might appear we can fix in a followup PR, makes sense?

**Comment by consideRatio** [comment]
I added an upgrade note in the README.md and wrote a more technical note in values.yaml about configJSON discussing YAML / JSON stuff.

**Comment by cpanato** [comment]
@consideRatio thanks so much for your work and time spend here.

**Review by cpanato — APPROVED** [review]
